Client/game_sa/CFileLoaderSA.cpp

```diff
@@ -159,3 +144,87 @@ RpAtomic* CFileLoader_SetRelatedModelInfoCB(RpAtomic* atomic, SRelatedModelInfo*
     }
     return atomic;
 }
+
+bool IsMatrixNeeed(SFileObjectInstance* pInst, CEntitySAInterface* pEntity)
+{
+    const bool bHasRotation = std::fabs(pInst->rotation.fX) > 0.050000001f || std::fabs(pInst->rotation.fY) > 0.050000001f ||
+        ((pInst->interiorID & 0xFF00) >> 8) & 2;
+    if (!bHasRotation) {
+        return false;
+    }
+
+    pInst->rotation.fX *= -1.0f;
+    pInst->rotation.fY *= -1.0f;
+    pInst->rotation.fZ *= -1.0f;
+
+    CPlaceableSAInterface* pPlaceable = reinterpret_cast<CPlaceableSAInterface*>(pEntity);
+
+    pPlaceable->AllocateStaticMatrix();
+
+    // We don't want to modify the original instance. So we have to copy the rotation.
+    CVector4D vecQuaternion(pInst->rotation);
+
+    /*
+        A quaternion is must be normalized. GTA is relying on an internal R* exporter and everything is OK,
+        but custom exporters might not contain the normalization. And we must do it yourself in MTA.
+    */
+    const float fLenSqr = vecQuaternion.LengthSquared();
+    if (std::fabs(fLenSqr - 1.0f) > std::numeric_limits<float>::epsilon() && fLenSqr > 0.0f)
+        vecQuaternion /= std::sqrt(fLenSqr);
+
+    if (!pEntity->Placeable.matrix) {
+        pPlaceable->AllocateMatrix();
+        pPlaceable->m_transform.UpdateMatrix(pEntity->Placeable.matrix);
+    }
+
+    pEntity->Placeable.matrix->SetRotate(vecQuaternion);
```

Comment:
This is inlined
`pPlaceable->GetMatrix()->SetRotate(vecQuaternion)`

see
https://github.com/codenulls/gta-reversed/blob/d5e83be9f188daf57902d0199f6c5019ba6f3e8d/source/game_sa/Entity/CPlaceable.cpp#L194

I've been also wondering if this issue could be related to max limit of 255 Collisions loaded in CColStore, see
https://github.com/codenulls/gta-reversed/blob/d5e83be9f188daf57902d0199f6c5019ba6f3e8d/source/game_sa/Collision/CColStore.cpp#L55

I've reversed most of the collision related code (other than the collision calculations themselves) there, it may be useful in looking for the solution

Reply:
Thanks, that's good to know.